src/engine/parser/error.rs

```diff
@@ -0,0 +1,190 @@
+// This file is part of rust-web/twig
+//
+// For the copyright and license information, please view the LICENSE
+// file that was distributed with this source code.
+
+//! Typisation of parser errors.
+
+use std::fmt::{self, Display};
+use error::Error;
+use error::{GeneralizeTo, ErrorCode};
+
+use engine::parser::token::TokenErrorCode;
+use engine::parser::job::{self, cursor};
+use engine::parser::token;
+
+pub type ParserError = Error<ParserErrorCode>;
+pub type NodeError = Error<NodeErrorCode>; // todo move somewhere else??
+
+impl GeneralizeTo<ParserErrorCode> for TokenErrorCode {
+    fn generalize(&self) -> ParserErrorCode { ParserErrorCode::TokenError }
+}
+
+#[allow(dead_code)]
+#[derive(Debug)]
+pub enum ParserErrorCode {
+    Unreachable {
+        reason: String,
+        job: job::JobDump,
+    },
+    MissingExtensions,
+    InvalidState {
```

Comment:
I know this is a **specific** case of "should not happen" alias `unreachable`.

But using error codes instead of panics in those weird places could really help debugging. The Lexer and Parser are just too complex too argue with 100% certainty that all "shouldn't be reachable" spots really are unreachable.

In my mind the list of error codes does not be a 1:1 list of "things that can go wrong if twig is used incorrectly" - I think it can also contain errors like "Oops, something went wrong".

Rust gives us two ways to Display those errors
* Display::fmt for production use
* Debug::fmt for debugging

I see nothing wrong to include some "He's dead Jimmy" errors with possible Debug-Info for us.

But again, I will remove it at least for now, since it is a detail.